cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,528 @@
+        """
+            Turns a numeric CPMpy 'flat' expression into a solver-specific
+            numeric expression
+        """
+
+        xcoefs = []
+        xvars = []
+        xrhs = 0
+        
+        assert is_num(rhs), "RHS of inequality should be numeric after transformations"
+        xrhs += rhs
+
+        if is_num(lhs):
+            xrhs -= lhs
+        elif isinstance(lhs, _NumVarImpl):
+            xcoefs += [1]
+            xvars += [self.solver_var(lhs)]
+        elif lhs.name == "sum":
+            xcoefs = [1]*len(lhs.args)
+            xvars = self.solver_vars(lhs.args)
+        elif lhs.name == "wsum":
+            xcoefs += lhs.args[0]
+            xvars += self.solver_vars(lhs.args[1])
+        else:
+            raise NotImplementedError("Exact: Unexpected lhs {} for expression {}".format(lhs.name,lhs))
+
+        return xcoefs,xvars,xrhs
+
+
+    def transform(self, cpm_expr):
+        """
+        Transform arbitrary CPMpy expressions to constraints the solver supports
+
+        Implemented through chaining multiple solver-independent **transformation functions** from
+        the `cpmpy/transformations/` directory.
+
+        See the 'Adding a new solver' docs on readthedocs for more information.
+
+        :param cpm_expr: CPMpy expression, or list thereof
+        :type cpm_expr: Expression or list of Expression
+
+        :return: list of Expression
+        """
+        # apply transformations, then post internally
+        # expressions have to be linearized to fit in MIP model. See /transformations/linearize
+        cpm_cons = flatten_constraint(cpm_expr)  # flat normal form
+        cpm_cons = decompose_global(cpm_cons, supported=frozenset({})) # alldiff has specialized MIP decomp in linearize
+        cpm_cons = reify_rewrite(cpm_cons, supported=frozenset(['sum', 'wsum']))  # constraints that support reification
+        cpm_cons = only_numexpr_equality(cpm_cons, supported=frozenset(["sum", "wsum"]))  # supports >, <, !=
+        cpm_cons = only_bv_implies(cpm_cons)  # anything that can create full reif should go above...
+        cpm_cons = linearize_constraint(cpm_cons, supported=frozenset({"sum","wsum"}))  # the core of the MIP-linearization
+        cpm_cons = only_positive_bv(cpm_cons)  # after linearisation, rewrite ~bv into 1-bv
+        return cpm_cons
+
+        # NOTE: the transformations that are still done specifically for Exact are two-fold:
+        # 1) transform '==' and '<=' to '>='
+        # 2) transform implications with negative conditions to ones with positive consequences
+        #
+        # 1) seems quite general and is a candidate to function as an independent transformation.
+        # 2) seems very solver-specific.
+
+    @staticmethod
+    def fix(o):
+        return o.item() if isinstance(o, np.generic) else o
+
+    def _add_xct_constr(self, xct_coefs,xct_vars,uselb,lb,useub,ub):
+        if any([not isinstance(x, numbers.Integral) for x in xct_coefs+[lb,ub]]):
```

Comment:
you can drop the `[]` in the `any`,  it should with generators as well.

Reply:
Same for `max` below

Reply:
done